coffee/web_socket_channel.coffee

```diff
@@ -30,18 +31,36 @@ class palava.WebSocketChannel extends @EventEmitter
       @emit 'open', handshake
     @socket.onmessage = (msg) =>
       try
-        @emit 'message', JSON.parse(msg.data)
+        parsedMsg = JSON.parse(msg.data)
+        if parsedMsg == {event: "pong"}
+          @outstandingPongs -= 1
+        else
+          @emit 'message', parsedMsg
       catch SyntaxError
         @emit 'error_invalid_json', msg
     @socket.onerror = (msg) =>
+      clearInterval(@pingInterval)
       if @retries > 0
         @retries -= 1
         @setupWebsocket()
+        @startClientPings()
       else
         @emit 'error', msg
     @socket.onclose = =>
+      clearInterval(@pingInterval)
       @emit 'close'
 
+  startClientPings: =>
+    @outstandingPongs = 0
+    @pingInterval = setInterval( () =>
+      if @outstandingPongs >= 6
+        clearInterval(@pingInterval)
+        @emit 'error', msg
```

Comment:
I'd love to see another event type here to prevent mix ups with low-level socket errors, maybe something like "no_pong" (also msg is not necessary)

Reply:
Let's discuss that a little. You're right that the it's a different symptom that we talk about here, but isn't it generally the same thing if the websocket throws an error (which it does on firefox quite quickly and on chrome it takes a looong time) or the pong does not get received - both show a (at least temporarily dead) websocket connection.

Reply:
in a first step I just exchanged "msg" by an error string now

Reply:
Sure. I agree, having one kind of semantic error event has some advantages. Still it feels not optimal to me, when sometimes the error object is an websocket-error-object and sometimes it's a string from our library

Reply:
One solution could be to have different error strings as first argument, like "socket_error", "no_pong". The more I think about it, the more I like it. We could also put "error_invalid_json" into an error type (the frontend does not listen to it, but it should. When the frontend listenes to the general "error" it would catch invalid jsons and put it through its default error dealing process)

Reply:
The actual socket error object would then be passed as 2nd parameter

Reply:
#23 implements this behavior